crates/sui/tests/cli_tests.rs

```diff
@@ -1876,7 +1876,10 @@ async fn test_package_management_on_upgrade_command() -> Result<(), anyhow::Erro
 
     // Create a new build config for the upgrade. Initialize its lock file
     // to the package we published.
-    let build_config_upgrade = BuildConfig::new_for_testing().config;
+    let mut build_config_upgrade = BuildConfig::new_for_testing().config;
```

Comment:
Ah, missed walking through this, but also tried handling this in a more principled way, considering your comment--tl;dr with long incoming thoughts:

TL;DR I think it's more a matter of `install_dir` getting in the way for publish/upgrade flows where we would never reasonable expect (or allow the user to put) a `Move.lock` other than the root directory for publishing/upgrading. Although we can (and I tried) defining a behavior around `install_dir` in these cases to be more hygienic for tests, it didn't pan out and needs more effort, where the benefit is dubious.

---

Thoughts:

- When we publish/upgrade, I don't believe there is a notion of a user-configurable `install_dir`, that's an option for `sui move build`. Which is to say, we currently expect the `Move.lock` to be in the root package dir, and `build/lock` workspace relative to the package root, when we publish/upgrade.

- As pointed out, the `install_dir` isn't being "respected" (if we decided to define respecting as meaning "we should access the lock file under `install_dir` always")
  - Though again, in this test, `install_dir` is set in a way that isn't possible via the CLI or for publishing AFAIK. If there were I might be more concerned about changing the inner values here, but with the current state the idea of `install_dir` being anything other than the the current directory on publish/upgrade is a kind of undefined behavior.

  - So why can't we just use this `install_dir` from `BuildConfig::new_from_testing` as-is and what would it take to "respect" it in publish/upgrade flows? 
    - I dug into this for quite some hours and felt like I got close to succeeding (didn't quite) and felt this is a big hassle for something that is only going to benefit a test setup scenario and not any realistic publish/upgrade flows. One reason is that when publishing and resolving IDs, [we use the package path](https://sourcegraph.com/github.com/MystenLabs/sui@cccd6409d90c9460d80892e83e6d18d72dc27f62/-/blob/crates/sui-package-management/src/lib.rs?L115), and would now need to use the custom-set `lock_path` in build config instead (and painfully propagate that through). Similar changes in this PR and for test setup.
      - Made lots of related changes and things still didn't work out (e.g., publish/upgrade actions still looking in the package dir for `Move.lock`), started questioning my sanity and if this is really worth it.

So, I think it's reasonable but unnecessary to use these values in `BuildConfig` for test purposes, but I'm fairly convinced that the operation of setting IDs in the `Move.lock` during publish/upgrade is disjoint from build configs/`install_dir` concerns, and okay to punt on needing that to work in this test.

Reply:
> When we publish/upgrade, I don't believe there is a notion of a user-configurable `install_dir`

I don't think this is true -- it is configurable using the `--install-dir` flag on `sui client publish` and `upgrade`. It's part of the `build_config: MoveBuildConfig` parameter that is flattened into the `Publish` and `Upgrade` commands.

Reply:
I believe the main reason why `install_dir` exists in `BuildConfig` is precisely for testing purposes -- so that when we write tests that require building Move packages in the repo, it doesn't dirty the repo state.

In CI this is especially important in the case of lock files because lock files rely on atomic rename to be durable, and that property only holds if the rename is within the same mount point. On CI, the temp directory uses tmpfs, so that's not true, and that is a major reason why `BuildConfig::new_for_testing()` currently exists.

The detail about us reading the lock file relative to the package path is an interesting one. My argument is that this does seem like the right thing to do for a dependency, because the only valid location for the lock file of a dependency is right next to its `Move.toml` (and in general we don't have a way to specify an install directory for dependencies, only the root package being built), but if we are using this same technique to look for the lock file of the root package, that's when we have a problem.

For me that means that if we're in this situation because our test requires setting up a dependency with its lock file and then using that in a future build, then the right thing to do is to construct our own `BuildConfig` without using a temp directory for the install dir (when building the dependency), as long as we are copying the dependency's package contents into a temp directory before doing so.

However, if we're in this situation because we are trying to read the root package's lock file from the wrong place then we should re-consider the implementation or test set-up.

Ultimately the thing that is strange about the code is to use the `new_for_testing` constructor but then undo the main thing that makes it different from the default `BuildConfig`.

Reply:
I see now that the dependency in this test is slightly different still -- it's to do with the lock file carrying version information between a publish and an upgrade of a package (in hindsight, it makes perfect sense that that's what we're talking about here).

To me this leans slightly more in the first camp -- that if we have been supplied an `install_dir` for a package, then that is where the upgrade command we should look for its lock file. I.e. the ideal way to write this test is to:

- Create a temp directory to act as the install dir.
- Build and publish the first version of the package, without copying it from the repo, but using the temp install dir.
- Build the second version of the package and upgrade it, again without copying it from the repo, but using the same build config, with the same install dir.

Reply:
> Sorry for the rambling thoughts about this whole BuildConfig thing

That's totally fine, useful context and thoughts, trying to figure out what's best here.

> I don't think this is true -- it is configurable using the `--install-dir`

TIL!! I was wrong, I stand corrected. Okay. So, after going through the thoughts and revisiting test setup, I still think the "this behavior is not user-configurable and not worth chasing a fix the mainline implementation". Because: AFAIK, there is actually no reason to care or honor installing `Move.lock` to `install_dir` (even if `install_dir` is user-configurable). AFAIK we only `lock.commit(somewhere)` where `somewhere` is what [the `lock_file` path is specified](https://sourcegraph.com/github.com/MystenLabs/sui@94a8236a61dbe90f1f9f8eea9364b12868670ff0/-/blob/external-crates/move/crates/move-package/src/lib.rs?L71-73). In the mainline execution we don't set `lock_file` relative to `install_dir` (we _do_ create `build/locks` relative to `install_dir`--the working space before committing locks). `lock_file` value is not a user-configurable value (this time I checked and we see clap skip).

All to say, the decision in the test setup to set the [`lock_file` path to `install_dir.join("Move.lock")`](https://sourcegraph.com/github.com/MystenLabs/sui@94a8236a61dbe90f1f9f8eea9364b12868670ff0/-/blob/crates/sui-move-build/src/lib.rs?L88) seems arbitrary wrt to `install_dir`, it could be any tmp dir. But presumably because that test setup conveniently creates a tmp dir for `install_dir`, it just sets the `lock_path` there as well. Point is, if a user configured `install-dir`, there isn't, AFAIK, a real need or intention to actually install `Move.lock` there for them--that's something this test setup uniquely decides to do, and presumably because the `Move.lock` can be checked for testing if it's output there.


Re suggestion:

> that if we have been supplied an install_dir for a package, then that is where the upgrade command we should look for its lock file.

So because of ^ I don't think it's the case that if we have an `install_dir` for a package, then we should look for the lock file there. In fact, when I realized that, it turns out I can remove the line that changes `install_dir` and keep the one in build config:

```patch
- build_config_upgrade.install_dir = Some(install_dir.clone());
```

And the test still passes. It's the case if we have a `lock_file` path (which is not user-configurable), that we should look for the lock file there. I.e., this line is important:

```patch
+ build_config_upgrade.lock_file = Some(pkg_root.join("Move.lock"));
```

To wit, and re: your suggesting--our mainline logic [wants to find the lock file in the source package root](https://sourcegraph.com/github.com/MystenLabs/sui@cccd6409d90c9460d80892e83e6d18d72dc27f62/-/blob/crates/sui-package-management/src/lib.rs?L115) not at the `lock_file` path (which may be overridden in some internal way, like in the test, but cannot realistically be overridden by a user otherwise). Changing the mainline code to always use `lock_file` path instead of "expect `Move.lock` in root of package" is the part I attempted to fix in the mainline code, turned out invasive, overly difficult, and eventually felt unnecessary for realistic use cases. That's why this test has the (perhaps not ideal) change to essentially say "no, just point `lock_path` to `Move.lock` at the package root").

I've updated the implementation in a way I think works a647c7f3fbcb8378d63250b4cc53c732a2f0e5ae. It is similar but simpler than before:

- Just one build config
- Keep `install_dir` in build config the way it is (it remains the working space for tmp lock files)
- Allow `publish` to put it's resulting `Move.lock` at the `BuildConfig` `lock_path` (i.e., `install_dir/Move.lock`)
- Set up the package for upgrading in a tmp directory (no clobbering) and copy the `install_dir/Move.lock` to the package root for upgrading.
- Point the build config `lock_file` to the `root/Move.lock` so it can be used as the input / output.
- Comment asserting that we simply don't support arbitrary `lock_file` in BuildConfig when that `lock_file` path needs to be used as an input (i.e., for upgrades). As explained above, supporting this `lock_file` to be arbitrary for the purposes of testing is really difficult (I tried). If we insist on asserting something more general re: arbitrary `lock_file` I can look at it in a follow up, but I won't lose sleep over assuming the `Move.lock` needs to be in a root directory for upgrades, and with these tests, we do that action in a tmp dir for the package to not clobber things.

Given my assumptions on "`Move.lock` in root" I think it would be fair to avoid the `BuildConfig::new_for_testing` and just use a sandbox for publish/upgrade where everything assumes the package root, but, like, the test achieves what it should at this point so `¯\_(ツ)_/¯`

Reply:
Thanks for digging in here! Main thing I'm taking away from this is that this space may need a general refactor, because this is too tough to navigate... 

> Given my assumptions on "`Move.lock` in root" I think it would be fair to avoid the `BuildConfig::new_for_testing` and just use a sandbox for publish/upgrade where everything assumes the package root, but, like, the test achieves what it should at this point so ¯\_(ツ)_/¯

Agreed -- simulating a "real" in-place upgrade where everything lives in a temp directory is probably the cleanest way to go (`new_for_testing` is really designed for the case where we're building a Move Package but don't care at all about the lock file or other artifacts -- just the in-memory representation of the package). But let's treat that as a follow-up.